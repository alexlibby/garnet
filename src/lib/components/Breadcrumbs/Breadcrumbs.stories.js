import Breadcrumbs from "./Breadcrumbs.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

let levels = [
  { href: "/", text: "Dashboard" },
  { href: "/reports", text: "Annual reports" },
  { href: "/reports/2019", text: "2019" },
];

const meta = {
  parameters: {
    badges: [],
  },
  argTypes: {
    BreadcrumbItems: levels,
    image: { control: "boolean" },
  },
};

export default {
  title: "Garnet UI Library/Navigation Components/Breadcrumbs",
  component: Breadcrumbs,
  ...meta,
};

export const Default = (args) => ({
  title: "Garnet UI Library/Navigation Components/Breadcrumbs",
  Component: Breadcrumbs,
  ...meta,
  props: {
    ...args,
    breadcrumbItems: levels,
  },
});

Default.parameters = {
  badges: [BADGES.ALPHA],
  docs: { source: { type: "code" } },
};

export const CustomImage = (args) => ({
  title: "Garnet UI Library/Navigation Components/Breadcrumbs",
  Component: Breadcrumbs,
  ...meta,
  props: {
    ...args,
    iconFamily: "material-symbols",
    iconName: "arrow-right-alt",
    breadcrumbItems: levels,
  },
});

CustomImage.parameters = {
  badges: [BADGES.ALPHA],
};

export const TextDivider = (args) => ({
  title: "Garnet UI Library/Navigation Components/Breadcrumbs",
  Component: Breadcrumbs,
  ...meta,
  props: {
    ...args,
    divider: ">>",
    breadcrumbItems: levels,
  },
});

TextDivider.parameters = {
  badges: [BADGES.ALPHA],
};
