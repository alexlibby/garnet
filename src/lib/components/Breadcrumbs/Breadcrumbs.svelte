<svelte:options customElement="garnet-breadcrumbs" /> 

<script>
  export let divider = "/";
  export let iconFamily = "";
  export let iconName = "";
  export let breadcrumbItems = [];

  import Icon from '@iconify/svelte';
</script>

<div class="garnet">
   <ul class="breadcrumb">
    {#each breadcrumbItems as breadcrumbIitem, i}
      <li>
        <!-- Breadcrumb divider -->
        {#if i !==0}
          {#if !iconFamily && !iconName} 
            <span>{divider}</span>
          {:else}
            <!-- Use icons from https://icon-sets.iconify.design/ -->
            <Icon icon={`${iconFamily}:${iconName}`} />
          {/if}
        {/if}

        <!-- Render each breadcrumb -->
        {#if i === breadcrumbItems.length - 1}
          {breadcrumbIitem.text}
        {:else}
          <a href={(breadcrumbIitem.href)} data-testid="breadcrumbLink">
            { breadcrumbIitem.text}
          </a>
        {/if}
      </li>
    {/each}
  </ul>
</div>

<style>
  .garnet { 
    display: flex; 
    font-family: Arial, Helvetica, sans-serif; 
  }
   
  ul.breadcrumb { 
    padding: 10px 16px; 
    list-style: none; 
    background-color: #eee; 
  }
  
  ul.breadcrumb li { 
    display: inline; 
    font-size: 18px; 
    vertical-align: text-bottom;
  } 
  
  ul.breadcrumb li a { 
    color: #733635; 
    text-decoration: none; 
    vertical-align: text-bottom;
  }

  ul.breadcrumb li a:hover {
    color: #d19c9b;
    text-decoration: underline;
  }

  ul.breadcrumb li span {
    display: inline;
    padding: 8px;
  }
</style>
