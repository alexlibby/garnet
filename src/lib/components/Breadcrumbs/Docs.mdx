import { Canvas, Meta } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as BreadcrumbsStories from "./Breadcrumbs.stories";

<Meta of={BreadcrumbsStories} />

# Breadcrumbs

Breadcrumbs is the primary component. It has three possible states.

- [Default](#default)
- [CustomImage](#customimage)
- [TextDivider](#textdivider)

<a id="default" />

## Default

This is the default version of the Breadcrumbs component.

<Canvas of={BreadcrumbsStories.Default} />

<a id="customimage" />

## CustomImage

This version of the Breadcrumbs component displays an icon from the Iconify library.

<Canvas of={BreadcrumbsStories.CustomImage} />

<a id="textdivider" />

## TextDivider

This version of the Breadcrumbs component displays a text-based divider.

<Canvas of={BreadcrumbsStories.TextDivider} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={BreadcrumbsStories} />
