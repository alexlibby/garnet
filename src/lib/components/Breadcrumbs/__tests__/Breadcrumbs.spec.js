import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Breadcrumbs from "../../Breadcrumbs/Breadcrumbs.svelte";

const BreadcrumbsHTML = `<div class="garnet s-A_BFCwAXJd84"><ul class="breadcrumb s-A_BFCwAXJd84"><li class="s-A_BFCwAXJd84"> <a href="/" data-testid="breadcrumbLink" class="s-A_BFCwAXJd84">Dashboard</a> </li><li class="s-A_BFCwAXJd84"><span class="s-A_BFCwAXJd84">/</span> <a href="/reports" data-testid="breadcrumbLink" class="s-A_BFCwAXJd84">Annual reports</a> </li><li class="s-A_BFCwAXJd84"><span class="s-A_BFCwAXJd84">/</span> 2019 </li></ul></div>`;
expect.extend(AxeMatchers);

describe("Tests for Breadcrumbs", () => {
  const props = {
    divider: "/",
    breadcrumbItems: [
      { href: "/", text: "Dashboard" },
      { href: "/reports", text: "Annual reports" },
      { href: "/reports/2019", text: "2019" },
    ],
  };

  test("should render properly", async () => {
    const result = render(Breadcrumbs, { props });
    expect(() => result).not.toThrow();
  });

  test("should show the Breadcrumbs when clicking the button", async () => {
    const result = render(Breadcrumbs, { props });
    const links = result.getAllByTestId("breadcrumbLink");
    expect(links.length).toBe(2);
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Breadcrumbs({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => BreadcrumbsHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
