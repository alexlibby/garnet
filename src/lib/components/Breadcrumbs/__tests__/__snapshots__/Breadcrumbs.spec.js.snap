// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Breadcrumbs > get a snapshot of component 1`] = `"<div class="garnet-breadcrumbs svelte-8qvfxh"><ul class="breadcrumb svelte-8qvfxh"><li class="svelte-8qvfxh"> <a href="/" data-testid="breadcrumbLink" class="svelte-8qvfxh">Dashboard</a> </li><li class="svelte-8qvfxh"><span class="svelte-8qvfxh">/</span> <a href="/reports" data-testid="breadcrumbLink" class="svelte-8qvfxh">Annual reports</a> </li><li class="svelte-8qvfxh"><span class="svelte-8qvfxh">/</span> 2019 </li></ul></div>"`;
